File: services/radarr.py
import logging

import requests

from services.config_manager import ConfigManager

logger = logging.getLogger(__name__)


class RadarrClient:
    def __init__(self):
        config = ConfigManager()
        self.host = config.get("RADARR_HOST")
        self.api_key = config.get("RADARR_API_KEY")

    def get_movies(self):
        if not self.host or not self.api_key:
            logger.warning("Radarr credentials not configured")
            return []

        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/movie"
            headers = {"X-Api-Key": self.api_key}

            response = requests.get(url, headers=headers, timeout=(5, 60))
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.error(f"Error fetching data from Radarr: {e}")
            return []

    def get_history(self, page_size=1000):
        if not self.host or not self.api_key:
            return []

        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/history"
            headers = {"X-Api-Key": self.api_key}
            params = {"pageSize": page_size}

            response = requests.get(
                url, headers=headers, params=params, timeout=(5, 60)
            )
            response.raise_for_status()
            return response.json().get("records", [])
        except Exception as e:
            logger.error(f"Error fetching history from Radarr: {e}")
            return []

    def delete_movie(self, movie_id):
        if not self.host or not self.api_key:
            return False

        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/movie/{movie_id}"
            headers = {"X-Api-Key": self.api_key}
            params = {"deleteFiles": "true"}

            response = requests.delete(
                url, headers=headers, params=params, timeout=(5, 60)
            )
            response.raise_for_status()
            logger.info(f"Deleted movie {movie_id} from Radarr.")
            return True
        except Exception as e:
            logger.error(f"Error deleting movie {movie_id} from Radarr: {e}")
            return False

    def get_disk_space(self):
        if not self.host or not self.api_key:
            return []

        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/diskspace"
            headers = {"X-Api-Key": self.api_key}

            response = requests.get(url, headers=headers, timeout=(5, 30))
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.error(f"Error fetching disk space from Radarr: {e}")
            return []

    def get_root_folders(self):
        if not self.host or not self.api_key:
            return []

        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/rootfolder"
            headers = {"X-Api-Key": self.api_key}

            response = requests.get(url, headers=headers, timeout=(5, 30))
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.error(f"Error fetching root folders from Radarr: {e}")
            return []

    def check_connection(self):
        if not self.host or not self.api_key:
            return False
        try:
            base_url = self.host.rstrip("/")
            url = f"{base_url}/api/v3/system/status"
            headers = {"X-Api-Key": self.api_key}
            requests.get(url, headers=headers, timeout=5).raise_for_status()
            return True
        except Exception:
            return False

File: src/backend/services/radarr.ts
import axios, { AxiosInstance } from 'axios';
import { ConfigManager } from './config_manager';

export class RadarrClient {
  private host: string | undefined;
  private apiKey: string | undefined;
  private api: AxiosInstance | null = null;

  constructor() {
    const config = new ConfigManager();
    this.host = config.get('RADARR_HOST');
    this.apiKey = config.get('RADARR_API_KEY');

    if (this.host && this.apiKey) {
      this.api = axios.create({
        baseURL: `${this.host.replace(/\/$/, '')}/api/v3`,
        headers: {
          'X-Api-Key': this.apiKey,
        },
        timeout: 60000,
      });
    }
  }

  async get_movies(): Promise<any[]> {
    if (!this.api) {
      console.warn('Radarr credentials not configured');
      return [];
    }
    try {
      const response = await this.api.get('/movie');
      return response.data;
    } catch (error) {
      console.error('Error fetching data from Radarr:', error);
      return [];
    }
  }

  async get_history(pageSize: number = 1000): Promise<any[]> {
    if (!this.api) return [];
    try {
      const response = await this.api.get('/history', {
        params: { pageSize },
      });
      return response.data.records || [];
    } catch (error) {
      console.error('Error fetching history from Radarr:', error);
      return [];
    }
  }

  async delete_movie(movieId: string | number): Promise<boolean> {
    if (!this.api) return false;
    try {
      await this.api.delete(`/movie/${movieId}`, {
        params: { deleteFiles: 'true' },
      });
      console.info(`Deleted movie ${movieId} from Radarr.`);
      return true;
    } catch (error) {
      console.error(`Error deleting movie ${movieId} from Radarr:`, error);
      return false;
    }
  }

  async get_disk_space(): Promise<any[]> {
    if (!this.api) return [];
    try {
      const response = await this.api.get('/diskspace');
      return response.data;
    } catch (error) {
      console.error('Error fetching disk space from Radarr:', error);
      return [];
    }
  }

  async get_root_folders(): Promise<any[]> {
    if (!this.api) return [];
    try {
      const response = await this.api.get('/rootfolder');
      return response.data;
    } catch (error) {
      console.error('Error fetching root folders from Radarr:', error);
      return [];
    }
  }

  async check_connection(): Promise<boolean> {
    if (!this.api) return false;
    try {
      await this.api.get('/system/status', { timeout: 5000 });
      return true;
    } catch (error) {
      return false;
    }
  }
}
